sql/sql_parse.cc

```diff
@@ -8375,7 +8375,7 @@ TABLE_LIST *st_select_lex::add_table_to_list(THD *thd,
   }
 
   bool has_alias_ptr= alias != nullptr;
-  void *memregion= thd->calloc(sizeof(TABLE_LIST));
```

Comment:
This `calloc` was present at line 8328 prior to the change, yet without crashes.  Is there something about the original code that prevents the compiler from optimizing away the memset/bfill?

Reply:
Can you please send along the other compiler flags that you mentioned in your commit message?  Thanks.

Reply:
>  Is there something about the original code that prevents the compiler from optimizing away the memset/bfill?

#### It used to work before https://github.com/MariaDB/server/commit/a8a75ba2d0dd639b2b14337a6c4f495fcaaff7e4 because it was using the C standard way to allocate memory and set variable values. 
The behavior was similar to following code https://gcc.godbolt.org/z/rWvcadcvb:
```
struct S {
  int i;
  S() {};
};

S* bar() {
  void *buf = malloc(sizeof(S));
  memset(buf, 0, sizeof(S));
  S* s = (S*) buf;     // later simply reads the allocated memory but no `new` function call
  return s;
}
```
which will be complied to `calloc` call (malloc + memset):
```
bar():
        mov     esi, 1
        mov     edi, 4
        jmp     calloc
```

#### With https://github.com/MariaDB/server/commit/a8a75ba2d0dd639b2b14337a6c4f495fcaaff7e4, the code uses the C++ `placement new` function. And the C++ `new` appears to rely on the constructor to initialize the memory contents of the new object. The compiler may optimize away any changes to the memory made before the constructor function is called.

Similar to following code https://gcc.godbolt.org/z/9rKhPc8s1:
```
struct S {
  int i;
  S() {};
};

S* bar() {
  void *buf = malloc(sizeof(S));
  memset(buf, 0, sizeof(S)); 
  S* s = new (buf) S;        // C++ way to initialize the object
  return s;
}
```
Which will be translate to a `malloc` call:
```
bar():
        mov     edi, 4
        jmp     malloc
```

About the optimization differences in GCC versions, GCC may argue that, rather than calling it a regression comparing GCC 4.9 to GCC 13, it's more like an optimizer enhancement leading to behavior change. As using C++ `new` to initialize the object means constructor function should handle the whole initialization so memory contents before `new` does not matter and is not guaranteed to be maintained, and using uninitialized variables of the object is still undefined behavior.

Reply:
> Can you please send along the other compiler flags that you mentioned in your commit message? Thanks.

There are lots of customization in CMake files in our build to adapt our internal build tools.

In general we are using GCC 13.2.0 and Link Time Optimization with -flto=auto. 

Some build flags from the build logs, when building sql_parse.cc.o:
```
-DHAVE_CONFIG_H -DHAVE_EVENT_SCHEDULER -DHAVE_OPENSSL -DHAVE_POOL_OF_THREADS -DMYSQL_SERVER -D_FILE_OFFSET_BITS=64 -IXXX -pthread -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -fdiagnostics-color=auto -fno-operator-names -D_GNU_SOURCE -flto=auto -pie -fPIC -fstack-protector --param=ssp-buffer-size=4 -O3 -march=core2 -mtune=generic -DNDEBUG -g -fgcse-after-reload -fipa-cp-clone -floop-interchange -floop-unroll-and-jam -fpeel-loops -fpredictive-commoning -fsplit-loops -fsplit-paths -ftree-loop-distribution -ftree-partial-pre -funswitch-loops -fvect-cost-model=dynamic -fversion-loops-for-strides -flto=auto -fno-omit-frame-pointer -D_FORTIFY_SOURCE=2 -DDBUG_OFF -Wall -Wenum-compare -Wenum-conversion -Wextra -Wmissing-braces -Wno-format-truncation -Wno-init-self -Wno-nonnull-compare -Wno-unused-parameter -Woverloaded-virtual -Wnon-virtual-dtor -Wvla -Wwrite-strings -std=gnu++11 -MD -MT
```
when linking mariadbd:
```
-pthread -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -fdiagnostics-color=auto -fno-operator-names -D_GNU_SOURCE -flto=auto -pie -fPIC -fstack-protector --param=ssp-buffer-size=4 -O3 -march=core2 -mtune=generic -DNDEBUG -g -O2 -fgcse-after-reload -fipa-cp-clone -floop-interchange -floop-unroll-and-jam -fpeel-loops -fpredictive-commoning -fsplit-loops -fsplit-paths -ftree-loop-distribution -ftree-partial-pre -funswitch-loops -fvect-cost-model=dynamic -fversion-loops-for-strides -flto=auto -fno-omit-frame-pointer -D_FORTIFY_SOURCE=2 -DDBUG_OFF -Wall -Wenum-compare -Wenum-conversion -Wextra -Wmissing-braces -Wno-format-truncation -Wno-init-self -Wno-nonnull-compare -Wno-unused-parameter -Woverloaded-virtual -Wnon-virtual-dtor -Wvla -Wwrite-strings -pthread -rdynamic -Wl,--as-needed -flto=auto -Wl,-z,relro,-z,now  -static-libstdc++ -fuse-ld=bfd  -Wl,--export-dynamic sql/CMakeFiles/mariadbd.dir/main.cc.o -o sql/mariadbd -LXXX -Wl,-rpath,XXX  sql/libsql.a  sql/libsql_builtins.a  LIB  vio/libvio.a  LIB  -lcrypt  sql/libwsrep.a  sql/libonline_alter_log.a  storage/maria/libaria.a  mysys_ssl/libmysys_ssl.a  LIB  LIB  sql/libpartition.a  storage/perfschema/libperfschema.a  sql/libsql_sequence.a  storage/csv/libcsv.a  storage/heap/libheap.a  storage/innobase/libinnobase.a  tpool/libtpool.a  LIB  LIB  storage/myisam/libmyisam.a  mysys/libmysys.a  dbug/libdbug.a  strings/libstrings.a  mysys/libmysys.a  dbug/libdbug.a  strings/libstrings.a  LIB  -lm  -ldl  storage/myisammrg/libmyisammrg.a  storage/sequence/libsequence.a  plugin/auth_socket/libauth_socket.a  plugin/provider_bzip2/libprovider_bzip2.a  -Wl,-Bstatic  -lbz2  plugin/provider_lz4/libprovider_lz4.a  -llz4  plugin/provider_lzma/libprovider_lzma.a  -llzma  plugin/provider_lzo/libprovider_lzo.a  -llzo2  plugin/rds_performance_insights/librds_performance_insights.a  plugin/type_geom/libtype_geom.a  plugin/type_inet/libtype_inet.a  plugin/type_uuid/libtype_uuid.a  plugin/user_variables/libuser_variables.a  plugin/userstat/libuserstat.a  sql/libthread_pool_info.a  -Wl,-Bdynamic
```